src/libs/ajax/User.ts

```diff
@@ -0,0 +1,398 @@
+import _ from 'lodash/fp';
+import * as qs from 'qs';
+import {
+  authOpts,
+  fetchBond,
+  fetchEcm,
+  fetchOk,
+  fetchOrchestration,
+  fetchRex,
+  fetchSam,
+  jsonBody,
+} from 'src/libs/ajax/ajax-common';
+import { getConfig } from 'src/libs/config';
+import { getTerraUser } from 'src/libs/state';
+import * as Utils from 'src/libs/utils';
+
+export interface SamUserRegistrationStatusResponse {
+  userSubjectId: string;
+  userEmail: string;
+  enabled: boolean;
+}
+
+export interface SamUserTosStatusResponse {
+  userInfo: {
+    userSubjectId: string;
+    userEmail: string;
+  };
+  enabled: {
+    ldap: boolean;
+    allUsersGroup: boolean;
+    google: boolean;
+  };
+}
+
+export interface SamUserTosComplianceStatusResponse {
+  userId: string;
+  userHasAcceptedLatestTos: boolean;
+  permitsSystemUsage: boolean;
+}
+
+export interface OrchestrationUserProfileResponse {
+  userId: string;
+  keyValuePairs: { key: any; value: any }[];
+}
+
+interface OrchestrationUserPreferLegacyFireCloudResponse {
+  preferTerra: boolean;
+  preferTerraLastUpdated: number;
+}
+
+interface NihDatasetPermission {
+  name: string;
+  authorized: boolean;
+}
+interface OrchestrationNihStatusResponse {
+  linkedNihUsername: string;
+  datasetPermissions: NihDatasetPermission[];
+  linkExpireTime: number;
+}
+
+interface BondFenceUrlResponse {
+  url: string;
+}
+
+interface BondFenceStatusResponse {
+  issued_at: Date;
+  username: string;
+}
+
+interface EcmLinkAccountResponse {
+  externalUserId: string;
+  expirationTimestamp: Date;
+  authenticated: boolean;
+}
+
+export interface SamInviteUserResponse {
+  userSubjectId: string;
+  userEmail: string;
+}
+interface RexFirstTimestampResponse {
+  timestamp: Date;
+}
+
+// TODO: Remove this as a part of https://broadworkbench.atlassian.net/browse/ID-460
+const getFirstTimeStamp = Utils.memoizeAsync(
+  async (token): Promise<RexFirstTimestampResponse> => {
+    const res = await fetchRex('firstTimestamps/record', _.mergeAll([authOpts(token), { method: 'POST' }]));
+    return res.json();
+  },
+  { keyFn: (...args) => JSON.stringify(args) }
+) as (...args: any[]) => RexFirstTimestampResponse;
+
+export const User = (signal?: AbortSignal) => {
+  return {
+    getStatus: async (): Promise<SamUserRegistrationStatusResponse> => {
+      const res = await fetchSam('register/user/v2/self/info', _.mergeAll([authOpts(), { signal }]));
+      return res.json();
+    },
+
+    profile: {
+      get: async (): Promise<OrchestrationUserProfileResponse> => {
+        const res = await fetchOrchestration('register/profile', _.merge(authOpts(), { signal }));
+        return res.json();
+      },
+
+      // We are not calling Thurloe directly because free credits logic was in orchestration
+      set: async (keysAndValues: {}): Promise<void> => {
```

Comment:
Is there a type for the input to `set` and `setPreferences`?

Reply:
I added a type for `set` which is just the `TerraUserProfile`. `setPreferences` is more tricky as it is storing notification settings per workspace, so the keys are something like `notifications/SuccessfulSubmissionNotification/${workspace.workspace.namespace}/${workspace.workspace.name}` with a boolean value. Other than that, it has the key `starredWorkspaces`. 

Not sure the best way to type the notification settings since the keys can be anything.